### PR TITLE
Refactor GenericMessageCorrelator & Add GenericMessageCorrelatorTest class

### PR DESCRIPTION
**Pull request by ChrisFlanTheMan**

### CODE REVIEW

**Review by LeoZ100 — APPROVED** [review]
Looks good to me.

File: messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/CorrelationData.java [hunk 1]
```diff
@@ -19,6 +29,37 @@ public CorrelationData(String messageType, String businessKey) {
...
+        String businessKey = evaluateExpression(documentContext, messageTypeExtensionData.getBusinessKeyExpression());
+        String tenantId = genericMessage.getTenantId();
+        boolean isStartEvent = messageTypeExtensionData.isStartEvent();
+        String processDefinitionKey = messageTypeExtensionData.getProcessDefinitionKey();
+
+        Map<String, String> matchVariables = createMatchVariablesFromExtensionData(documentContext, messageTypeExtensionData.getMatchVariableExpressions());
+        Map<String, String> inputVariables = createInputVariablesFromExtensionData(documentContext, messageTypeExtensionData.getInputVariableExpressions());
+
+        return new CorrelationData(messageType, tenantId, businessKey, processDefinitionKey, isStartEvent, matchVariables, inputVariables);
+    }
```
Comment (dare2try) [inline messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/CorrelationData.java:61]: I'm not really a fan of this change... this class was originally just a POJO but now has some pretty suffocated initialization logic

File: messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/CorrelationData.java [hunk 2]
```diff
@@ -35,6 +76,10 @@ public void setTenantId(String value) {
         this.tenantId = value;
     }
 
+    public boolean hasNonZeroTenantId() {
+        return !ZERO_UUID.equals(tenantId);
+    }
```
Comment (dare2try) [inline messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/CorrelationData.java:81]: I would rather have this outside of this class

File: messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/GenericMessageCorrelationBuilder.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    public MessageCorrelationResult correlate() {
+        try{
+            return messageCorrelationBuilder.correlateWithResult();
+        } catch (Exception ex) {
+            LOGGER.warning(ex.toString());
+            throw ex;
+        }
+    }
```
Comment (dare2try) [inline messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/GenericMessageCorrelationBuilder.java:58]: These two methods have nothing to do with your builder

File: messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/GenericMessageCorrelator.java [hunk 4]
```diff
@@ -31,135 +23,62 @@ public GenericMessageCorrelator(RuntimeService runtimeService) {
...
-        });
-
-        return results;
+        return determineCorrelatableProcessesInstances(correlationData)
+            .map((execution -> executeCatchMessageEventCorrelationByExecution(correlationData, execution)))
+            .collect(Collectors.toList());
     }
 
-    private List<Execution> determineCorrelatableProcessesInstances(CorrelationData correlationData) {
+    private Stream<Execution> determineCorrelatableProcessesInstances(CorrelationData correlationData) {
```
Comment (dare2try) [inline messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/GenericMessageCorrelator.java:61]: What is the purpose of this change?

Comment (dare2try) [reply messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/GenericMessageCorrelator.java:61]: I see why you changed it to `stream`, you can disregard

**Review by dare2try — CHANGES_REQUESTED** [review]

File: messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/CorrelationData.java (hunk 1 above)
Comment (ChrisFlanTheMan) [reply messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/CorrelationData.java:61]: Do you have a better suggestion for where to put it, like a CorrelationDataFactory class? In my opinion, initialization logic for a CorrelationData class should live as close to the class as possible, and not be the responsibility of the GenericMessageCorrelator to create. Keeping the GenericMessageCorrelator class with only business logic will help make it clean and readable, and most users will not care about the specifics of how this POJO is created.

File: messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/CorrelationData.java (hunk 2 above)
Comment (ChrisFlanTheMan) [reply messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/CorrelationData.java:81]: why? Can you give me a better example of where to put it? The GenericMessageCorrelator already has enough other stuff to worry about, so why not let a class like this control what a non zero tenant id means, since its always used in reference to this?

File: messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/GenericMessageCorrelationBuilder.java (hunk 3 above)
Comment (ChrisFlanTheMan) [reply messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/GenericMessageCorrelationBuilder.java:58]: This builder is effectively a wrapper class around MessageCorrelationBuilder. Similar to my other comments, where else would you want this to live outside of the GenericMessageCorrelator?

File: messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/CorrelationData.java (hunk 1 above)
Comment (dare2try) [reply messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/CorrelationData.java:61]: I would probably start off with something like `CorrelationUtils.buildCorrelationData(...)`, but in general I am opposed to your premise that the ways a POJO can be initialized should be co-located with the POJO itself.  A POJO should at minimum define the required set of attributes a particular domain object is responsible for, and at maximum provide helper functions against the attributes it is responsible for.  Other classes (e.g. builders, or utility classes) should be responsible for constructing these objects in various ways so that they CAN be constructed in various ways.

File: messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/CorrelationData.java (hunk 2 above)
Comment (dare2try) [reply messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/CorrelationData.java:81]: This would probably be a good candidate for a `TenantUtils.IsZeroTenant(...)` for instance

File: messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/GenericMessageCorrelationBuilder.java (hunk 3 above)
Comment (dare2try) [reply messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/GenericMessageCorrelationBuilder.java:58]: Now that I review this class again, I see all it's doing is proxying/delegating the calls to the `messageCorrelationBuilder` and I have to ask myself, what is the purpose of this class at all

Comment (ChrisFlanTheMan) [reply messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/GenericMessageCorrelationBuilder.java:58]: Good point, will see how this can be improved.

Comment (ChrisFlanTheMan) [reply messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/GenericMessageCorrelationBuilder.java]: Upon reviewing that class, I think i wrote it while trying to clean up the code, before i knew what the end state would look like. Removing it seems fine and doesn't add any more complexion to the GenericMessageCorrelator class

File: messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/TenantUtils.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package com.ultimatesoftware.workflow.messaging.correlation;
+
+import static com.ultimatesoftware.workflow.messaging.Constants.ZERO_UUID;
+
+public final class TenantUtils {
+
+    private TenantUtils() {}
+
+    public static boolean isNonZeroTenantId(String tenantId) {
```
Comment (dare2try) [inline messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/TenantUtils.java:9]: Can we just create the function `isSystemTenant` and then you can just negate that when you need to?

**Review by dare2try — CHANGES_REQUESTED** [review]

File: messaging/src/test/java/com/ultimatesoftware/workflow/messaging/builders/GenericMessageBuilder.java [hunk 6]
```diff
@@ -1,19 +1,35 @@
...
     private final String schemaVersion = "1.0";
-    private final String tenantId = GENERIC_TENANT_ID;
+    private String tenantId = GENERIC_TENANT_ID;
     private final String messageType = GENERIC_MESSAGE_TYPE;
-    private JsonNode body;
+    private JsonNode body = new ObjectMapper().createObjectNode()
+        .put(GENERIC_BUSINESS_PROCESS_KEY_FIELD, GENERIC_BUSINESS_PROCESS_KEY_VALUE)
+        .put("name", "name");
+
+    public GenericMessageBuilder withZeroTenantId() {
```
Comment (dare2try) [inline messaging/src/test/java/com/ultimatesoftware/workflow/messaging/builders/GenericMessageBuilder.java:24]: You might want to change this to `withSystemTenantId()` now 😅

File: messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/TenantUtils.java [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+import static com.ultimatesoftware.workflow.messaging.Constants.ZERO_UUID;
+
+public final class TenantUtils {
+
+    private TenantUtils() {}
+
+    public static boolean isSystemTenant(String tenantId) {
+        return !ZERO_UUID.equals(tenantId);
+    }
```
Comment (dare2try) [inline messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/TenantUtils.java:11]: Your return value is incorrect based on the method

**Review by dare2try — APPROVED** [review]
Thank you for all the changes!
